src/Data/List/Ext/Properties.agda

```diff
@@ -54,11 +161,382 @@ module _ {a} {A : Set a} ⦃ _ : DecEq A ⦄ where
+  x∈Allfm∘insertx {[]} = All.[]
+  x∈Allfm∘insertx {_ ∷ ls} = All++intro x∈Allinsertx (x∈Allfm∘insertx{ls})
+
+  l∈fm∘insertx→x∈l : {ls : List (List A)}{x : A} → ∀ l → l ∈ concatMap (insert x) ls → x ∈ l
+  l∈fm∘insertx→x∈l {ls} = ∈All-def (x∈Allfm∘insertx{ls})
+
+  x∈All→x∈Allfm∘insert : {ls : List (List A)}{y x : A} → All (x ∈_) ls → All (x ∈_) (concatMap (insert y) ls)
+  x∈All→x∈Allfm∘insert {.[]} All.[] = All.[]
+  x∈All→x∈Allfm∘insert {.(_ ∷ _)} (px All.∷ p) = All++intro (x∈→x∈Allinsert px) (x∈All→x∈Allfm∘insert p)
+
+  l∈insert→l⊆∷ : {xs : List A}{x : A} → ∀ l → l ∈ (insert x xs) → l ⊆ (x ∷ xs)
+  l∈insert→l⊆∷ {[]} {x} .(x ∷ []) (here refl) {z} z∈l = z∈l
+  l∈insert→l⊆∷ {x' ∷ xs} {x} .(x ∷ x' ∷ xs) (here refl) {z} z∈l = z∈l
+  l∈insert→l⊆∷ {x' ∷ xs} {x} (y ∷ ys) (there l∈) {.y} (here refl) = there (here (proj₁ (map∷decomp∈ l∈)))
+  l∈insert→l⊆∷ {x' ∷ xs} {x} (y ∷ ys) (there l∈) {z} (there z∈yys) with (l∈insert→l⊆∷ ys (proj₂ (map∷decomp∈ l∈)) z∈yys)
+  ...| here px = here px
+  ...| there pxs = there (there pxs)
+
+  l∈map∷→l∈fm∘insert : {ls : List (List A)}{y : A} → ∀ l → l ∈ map (y ∷_) ls → l ∈ concatMap (insert y) ls
+  l∈map∷→l∈fm∘insert {[] ∷ ls} {y} l (here px) = here px
+  l∈map∷→l∈fm∘insert {(_ ∷ _) ∷ _} l (here px) = here px
+  l∈map∷→l∈fm∘insert {[] ∷ _} l (there l∈) = there (l∈map∷→l∈fm∘insert l l∈)
+  l∈map∷→l∈fm∘insert {(_ ∷ _) ∷ _} l (there l∈) = xs⊆ys++xs _ _ (l∈map∷→l∈fm∘insert l l∈)
+
+  l∈map³→l∈fm∘insert∘map² : {ls : List (List A)}{y' y z : A} → ∀ l
+                            → l ∈ map (y' ∷_) (map (y ∷_) (map (z ∷_) ls))
+                            → l ∈ concatMap (insert y) (map (y' ∷_) (map (z ∷_) ls))
+  l∈map³→l∈fm∘insert∘map² {_ ∷ _} l (here px) = there (here px)
+  l∈map³→l∈fm∘insert∘map² {_ ∷ _} l (there l∈) = xs⊆ys++xs _ _ (l∈map³→l∈fm∘insert∘map² l l∈)
+
+  insert-decomp : {ls : List (List A)}{x : A} → ∀ ys → ys ∈ (concatMap (insert x) ls)
+                  → ∃[ l ] (l ∈ ls × ys ⊆ (x ∷ l))
+  insert-decomp {xs ∷ _} ys ys∈ with ++→⊎ ys∈
+  ...| inj₁ v = xs , here refl , l∈insert→l⊆∷ ys v
+  ...| inj₂ v with insert-decomp ys v
+  ...| l , l∈ls , ys⊆ = l , there l∈ls , ys⊆
+
+  insert-decomp∈ : {ls : List (List A)}{y : A} → ∀ xs → xs ∈ (concatMap (insert y) ls)
+                   → ∃[ l ] (l ∈ ls × xs ∈ (insert y l))
+  insert-decomp∈ {zs ∷ _} xs xs∈ with ++→⊎ xs∈
+  ...| inj₁ v = zs , here refl , v
+  ...| inj₂ v with insert-decomp∈ xs v
+  ...| l , l∈ls , xs∈yl = l , there l∈ls , xs∈yl
+
+  insert-decomp⁻¹ : {ls : List (List A)}{y : A} → ∀ xs → ∃[ l ] (l ∈ ls × xs ∈ (insert y l))
+                         → xs ∈ (concatMap (insert y) ls)
+  insert-decomp⁻¹ {l ∷ ls} {y} xs (.l , here refl , xs∈) = xs⊆xs++ys _ _ xs∈
+  insert-decomp⁻¹ {l ∷ ls} {y} xs (l' , there l'∈ , xs∈) = xs⊆ys++xs _ _ (insert-decomp⁻¹ xs (l' , (l'∈ , xs∈)))
+
+  insert++distr : {ll lr : List (List A)}{xs : List A}{x : A} → xs ∈ concatMap (insert x) (ll ++ lr)
+                     → xs ∈ concatMap (insert x) ll ++ concatMap (insert x) lr
```

Comment:
This proof and the ones below get way easier if you have
```
concatMap-++ : {f : A → List B} {l l' : List A} → concatMap f (l ++ l') ≡ concatMap f l ++ concatMap f l'
```

Reply:
https://omelkonian.github.io/formal-prelude/Prelude.Lists.Concat.html#701

Reply:
Yes, good points.  I'll try to use them to simplify things.

By the way, as I realized just yesterday, I could probably clean up a bunch of stuff by using the following minor embellishment of the library's `concat⁺`:

```agda
  concatMap⁺ : {ys xs : List A}{f : A → List A} → xs ⊆ ys → concatMap f xs ⊆ concatMap f ys
  concatMap⁺ = concat⁺ ∘ (map⁺ _)
```

Reply:
Indeed, the following should help simplify things:

```agda
  concatMap⁺ : {ys xs : List A}{f : A → List A} → xs ⊆ ys → concatMap f xs ⊆ concatMap f ys
  concatMap⁺ = concat⁺ ∘ (map⁺ _)

  concatMap-++ : ∀{b}{B : Set b} → ∀ (f : A → List B) (xs ys : List A)
    → concatMap f (xs ++ ys) ≡ concatMap f xs ++ concatMap f ys
  concatMap-++ f xs ys = begin
    concatMap f (xs ++ ys)           ≡⟨⟩
    concat (map f (xs ++ ys))        ≡⟨ cong concat $ map-++ f xs ys ⟩
    concat (map f xs ++ map f ys)    ≡˘⟨ concat-++ (map f xs) (map f ys) ⟩
    concatMap f xs ++ concatMap f ys ∎ where open ≡-Reasoning
```

The former is the one I mentioned above.  The latter is from the link @omelkonian shared.  Thanks Orestis!